File: backend/app/services/queue.py
"""
Queue utilities
"""
from typing import TypeVar, Generic, List, Optional
from collections import deque

T = TypeVar('T')


class Queue(Generic[T]):
    """Simple queue implementation"""
    def __init__(self):
        self._items: deque[T] = deque()

    def enqueue(self, item: T) -> None:
        """Add item to queue"""
        self._items.append(item)

    def dequeue(self) -> Optional[T]:
        """Remove and return item from queue"""
        if self._items:
            return self._items.popleft()
        return None

    def peek(self) -> Optional[T]:
        """Return first item without removing"""
        if self._items:
            return self._items[0]
        return None

    @property
    def size(self) -> int:
        """Get queue size"""
        return len(self._items)

    @property
    def is_empty(self) -> bool:
        """Check if queue is empty"""
        return len(self._items) == 0

    def clear(self) -> None:
        """Clear all items"""
        self._items.clear()

    def to_list(self) -> List[T]:
        """Convert to list"""
        return list(self._items)


class PriorityQueue(Generic[T]):
    """Priority queue implementation"""
    def __init__(self):
        self._items: List[tuple[int, T]] = []

    def enqueue(self, item: T, priority: int) -> None:
        """Add item with priority"""
        self._items.append((priority, item))
        self._items.sort(key=lambda x: x[0])

    def dequeue(self) -> Optional[T]:
        """Remove and return highest priority item"""
        if self._items:
            return self._items.pop(0)[1]
        return None

    def peek(self) -> Optional[T]:
        """Return highest priority item without removing"""
        if self._items:
            return self._items[0][1]
        return None

    @property
    def size(self) -> int:
        """Get queue size"""
        return len(self._items)

    @property
    def is_empty(self) -> bool:
        """Check if queue is empty"""
        return len(self._items) == 0

    def clear(self) -> None:
        """Clear all items"""
        self._items.clear()

File: src/lib/queue.ts
/**
 * Queue utilities
 */

export class Queue<T> {
  private items: T[] = []

  enqueue(item: T): void {
    this.items.push(item)
  }

  dequeue(): T | undefined {
    return this.items.shift()
  }

  peek(): T | undefined {
    return this.items[0]
  }

  get size(): number {
    return this.items.length
  }

  get isEmpty(): boolean {
    return this.items.length === 0
  }

  clear(): void {
    this.items = []
  }

  toArray(): T[] {
    return [...this.items]
  }
}

export class PriorityQueue<T> {
  private items: { item: T; priority: number }[] = []

  enqueue(item: T, priority: number): void {
    const element = { item, priority }
    let added = false

    for (let i = 0; i < this.items.length; i++) {
      if (this.items[i].priority > priority) {
        this.items.splice(i, 0, element)
        added = true
        break
      }
    }

    if (!added) {
      this.items.push(element)
    }
  }

  dequeue(): T | undefined {
    return this.items.shift()?.item
  }

  peek(): T | undefined {
    return this.items[0]?.item
  }

  get size(): number {
    return this.items.length
  }

  get isEmpty(): boolean {
    return this.items.length === 0
  }

  clear(): void {
    this.items = []
  }
}
